builtin/aws/ecs/platform.go

```diff
@@ -119,650 +133,845 @@ func (p *Platform) ValidateAuth() error {
-			if err != nil {
-				return err
-			}
-
-			return nil
-		},
-
-		Run: func(s LifecycleStatus) error {
-			dep, err = p.Launch(ctx, s, log, ui, sess, src, img, deployConfig, executionRole, taskRole, cluster, logGroup)
-			return err
-		},
-
-		Cleanup: func(s LifecycleStatus) error { return nil },
-	}
-
-	if err := lf.Execute(log, ui); err != nil {
-		return nil, err
-	}
-
-	return dep, nil
-}
-
-func defaultSubnets(ctx context.Context, sess *session.Session) ([]*string, error) {
-	svc := ec2.New(sess)
-
-	desc, err := svc.DescribeSubnets(&ec2.DescribeSubnetsInput{
-		Filters: []*ec2.Filter{
-			{
-				Name:   aws.String("default-for-az"),
-				Values: []*string{aws.String("true")},
-			},
-		},
-	})
+	// Destroy
+	err := rm.DestroyAll(ctx, log, sg, ui)
 	if err != nil {
-		return nil, err
-	}
-
-	var subnets []*string
-
-	for _, subnet := range desc.Subnets {
-		subnets = append(subnets, subnet.SubnetId)
+		return status.Errorf(codes.Internal, "failed to destroy all resources for deployment: %s", err)
 	}
 
-	return subnets, nil
+	s.Update("Finished destroying ECS deployment")
+	s.Done()
+	return nil
 }
 
-func (p *Platform) SetupCluster(ctx context.Context, s LifecycleStatus, sess *session.Session) (string, error) {
-	ecsSvc := ecs.New(sess)
+func (p *Platform) resourceClusterCreate(
+	ctx context.Context,
+	sg terminal.StepGroup,
+	log hclog.Logger,
+	sess *session.Session,
+	state *Resource_Cluster,
+) error {
+	s := sg.Add("Initiating cluster creation...")
+	defer s.Abort()
 
 	cluster := p.config.Cluster
 	if cluster == "" {
 		cluster = "waypoint"
 	}
+	state.Name = cluster
 
-	desc, err := ecsSvc.DescribeClusters(&ecs.DescribeClustersInput{
+	s.Update("Attempting to find existing cluster named %q", cluster)
+
+	ecsSvc := ecs.New(sess)
+	desc, err := ecsSvc.DescribeClustersWithContext(ctx, &ecs.DescribeClustersInput{
 		Clusters: []*string{aws.String(cluster)},
 	})
 	if err != nil {
-		return "", err
+		return err
 	}
 
 	for _, c := range desc.Clusters {
-		if *c.ClusterName == cluster && strings.ToLower(*c.Status) == "active" {
-			s.Status("Found existing ECS cluster: %s", cluster)
-			return cluster, nil
+		if *c.ClusterName == cluster {
+			if *c.Status == "PROVISIONING" {
+				s.Update("Existing ecs cluster %q is still provisioning - try again later.", cluster)
+			} else if *c.Status == "ACTIVE" {
+				s.Update("Using existing ECS cluster %s", cluster)
+				if c.ClusterArn != nil {
+					state.Arn = *c.ClusterArn
+				}
+				s.Done()
+				return nil
+			} else {
+				// Warn if we encounter waypoint clusters in other odd states (i.e. DEPROVISIONING, FAILED, etc.)
+				// I think it's ok to try to create a new cluster if one exists in a non-active non-provisioning state
+				log.Warn("Ignoring cluster named %q in state %q", cluster, *c.Status)
+			}
 		}
 	}
 
 	if p.config.EC2Cluster {
-		return "", fmt.Errorf("EC2 clusters can not be automatically created")
+		return fmt.Errorf("EC2 clusters can not be automatically created")
```

Comment:
Since we're changing this in this PR....we should probably opt for the `status.Error` + `codes.CodeID` everywhere here when we're crafting our own error (Versus just adding some extra text and just returning an error we received).

Reply:
I actually started by doing that! But I realized that these errors will return to the top-level Deploy function, which will wrap it in `status.Errorf` before the plugin returns.